src/javascript/track.js

```diff
@@ -7,39 +7,116 @@ export default class DropTrack {
     this.x = x;
     this.width = width;
     this.keyBind = keyBind;
-    this.particleEffect = new HitParticleEffect(vm.ctx);
+    this.particleEffect = new HitEffect(vm.ctx);
     this.noteArr = [];
     this.hitIndicatorOpacity = 0;
+    this.isKeyDown = false;
+    this.isUserHoldingNote = false;
     this.updateHitGradient();
   }
 
-  keyDown(key, color) {
-    if (key.includes(this.keyBind)) {
-      if (color !== "grey") this.hitIndicatorOpacity = 1;
-      if (!this.vm.playMode) {
-        // create mode
-        this.noteArr.push(new Note(this.vm, this.x, this.width, color));
-      } else if (this.noteArr && this.noteArr[0]) {
+  keyDown(key) {
+    if (key.includes(this.keyBind) && !this.isKeyDown) {
+      this.isKeyDown = true;
+      this.hitIndicatorOpacity = 1;
+      const createParticle = () => {
+        this.particleEffect.create(
+          this.x,
+          this.vm.checkHitLineY,
+          this.width,
+          10,
+          this.vm.markJudge
+        );
+      };
+      if (this.vm.playMode && this.noteArr && this.noteArr[0]) {
         const noteToDismiss = this.noteArr[0];
         if (noteToDismiss.getDiffPercentage() < 0.5) {
-          noteToDismiss.hitAndCountScore();
-          this.noteArr.shift();
-          this.particleEffect.create(
-            this.x,
-            this.vm.checkHitLineY,
-            this.width,
-            10
-          );
+          noteToDismiss.calculatePercent();
+          if (noteToDismiss.isHoldNote) {
+            if (noteToDismiss.noteFailed) return;
+            this.isUserHoldingNote = true;
+            let countInterval = setInterval(() => {
+              // if hold note is finished or is nearly finished but player released key, count as success
+              if (
+                noteToDismiss.isHoldNoteFinished() ||
+                (noteToDismiss.isHoldNoteFinished(true) &&
```

Comment:
Wait... what? How is `isHoldNoteFinished` variadic over 0/1 argument(s)?

Reply:
Should really be `false` in the first one, indicating `checking note is strictly finished`, I'll add it now